### PR TITLE
providers/aws: Add ElastiCache replication group support (fixes #1799)

### PR DESCRIPTION
**Pull request by saulshanabrook**
This is based off of @inferiorhumanorgans work in #1799. I fixed some import paths and added a `primary_endpoint` attribute, so it can be used like this:

```
resource "aws_elasticache_replication_group" "default" {
    replication_group_id = "redis.fun"

    description = "redis fun"
    engine = "redis"
    cache_node_type = "${var.node_type}"
    num_cache_clusters = 2
    automatic_failover = true
    subnet_group_name = "${var.subnet_group_name}"
    security_group_ids  = ["${var.security_group_id}"]
}

resource "aws_route53_record" "default" {
    zone_id = "${var.zone_id}"
    name = "${var.engine}-${var.name}.${var.zone_name}"
    type = "CNAME"
    ttl = "30"
    records = [
        "${aws_elasticache_replication_group.default.primary_endpoint}."
    ]
}
```

### CODE REVIEW

**Comment by saulshanabrook** [comment]
Looks like my tests failed:

```
builtin/providers/aws/resource_aws_elasticache_replication_group.go:109: undefined: aws.Boolean
builtin/providers/aws/resource_aws_elasticache_replication_group.go:110: undefined: aws.Long
builtin/providers/aws/resource_aws_elasticache_replication_group.go:182: undefined: aws.Boolean
builtin/providers/aws/resource_aws_elasticache_replication_group.go:188: undefined: aws.Boolean
```

Weird. It builds and passes tests on my local machine.

**Comment by dlsniper** [comment]
It might be that you use a different version of the aws sdk than the project, check it out: https://github.com/hashicorp/terraform/blob/master/deps/v0-6-1.json#L42 . Hope this helps.

**Comment by saulshanabrook** [comment]
@dlsniper Those deps look like they are for the 0.6.1 release. Do you know how I update the deps for the next release?

**Comment by dlsniper** [comment]
Sorry, that's something I can't know, maybe someone else does? I've noticed some problems when running against the latest AWS sdk as well so I'm in the dark as well.

**Comment by joshgarnett** [comment]
@saulshanabrook Looking at http://docs.aws.amazon.com/sdk-for-go/api/aws.html I don't see aws.Boolean or aws.Long.  It looks like there was a breaking change introduced in the SDK.  Take a look at https://github.com/hashicorp/terraform/commit/579ccbefea4ae52bfbccb6e643481e0b5ceb4230 for an example where this was fixed in the other resources.

**Comment by saulshanabrook** [comment]
@josephholsten That makes sense. OK I am trying to do `make updatedeps` but it keeps failing:

```
	imports google.golang.org/api/compute/v0.beta: cannot find package "google.golang.org/api/compute/v0.beta" in any of:
	/usr/local/Cellar/go/1.4.2/libexec/src/google.golang.org/api/compute/v0.beta (from $GOROOT)
	/Users/saul/projects/go/src/google.golang.org/api/compute/v0.beta (from $GOPATH)
```

**Comment by dlsniper** [comment]
Seems there's that it has been removed in this commit: https://github.com/google/google-api-go-client/commit/18450f4e95c7e76ce3a5dc3a8cb7178ab6d56121 (warning, it's a bit big). This looks like the new version: https://godoc.org/google.golang.org/api/compute/v1

**Comment by saulshanabrook** [comment]
OK I am trying to rebase off of the `master` branch and then run `make updatedeps` again to see if that solves it

**Comment by saulshanabrook** [comment]
OK updated with new names.

**Comment by josephholsten** [comment]
@saulshanabrook pretty sure you meant @joshgarnett, but I could definitely use this stuff. Thanks!

**Comment by joshgarnett** [comment]
@catsby any chance this can get merged soon?

**Comment by blacked** [comment]
It works in general. But AZ distribution works weird. It put 2 boxes in same AZ, however when I create cluster witn AWS Console, it put them in different AZ. 

```
resource "aws_elasticache_replication_group" "redis" {
    replication_group_id = "${concat(var.epop, \"-redis\")}"
    description = "cep-sjc-redis"
    engine = "redis"
    cache_node_type = "cache.m3.medium"
    num_cache_clusters = 2
    automatic_failover = true
    subnet_group_name = "${aws_elasticache_subnet_group.redis.name}"
    security_group_ids = ["${aws_security_group.redis.id}"]
    depends_on = ["aws_elasticache_subnet_group.redis"]
}
```
Result:
```
{
    "ReplicationGroups": [
        {
            "Status": "available",
            "Description": "test-sjc-redis",
            "NodeGroups": [
                {
                    "Status": "available",
                    "NodeGroupMembers": [
                        {
                            "CurrentRole": "primary",
                            "PreferredAvailabilityZone": "us-west-1c",
                            "CacheNodeId": "0001",
                            "ReadEndpoint": {
                                "Port": 6379,
                                "Address": "test-sjc-redis-001.lcuh0x.0001.usw1.cache.amazonaws.com"
                            },
                            "CacheClusterId": "test-sjc-redis-001"
                        },
                        {
                            "CurrentRole": "replica",
                            "PreferredAvailabilityZone": "us-west-1c",
                            "CacheNodeId": "0001",
                            "ReadEndpoint": {
                                "Port": 6379,
                                "Address": "test-sjc-redis-002.lcuh0x.0001.usw1.cache.amazonaws.com"
                            },
                            "CacheClusterId": "test-sjc-redis-002"
                        }
                    ],
                    "NodeGroupId": "0001",
                    "PrimaryEndpoint": {
                        "Port": 6379,
                        "Address": "test-sjc-redis.lcuh0x.ng.0001.usw1.cache.amazonaws.com"
                    }
                }
            ],
            "ReplicationGroupId": "test-sjc-redis",
            "AutomaticFailover": "enabled",
            "MemberClusters": [
                "test-sjc-redis-001",
                "test-sjc-redis-002"
            ],
            "PendingModifiedValues": {}
        }
    ]
}
```

**Comment by blacked** [comment]
@saulshanabrook 
I made [PR with this feature support](https://github.com/saulshanabrook/terraform/pull/2) to your fork.

**Comment by saulshanabrook** [comment]
Thanks @blacked, merged. This was his example usage, from his PR:


```
resource "aws_elasticache_replication_group" "redis" {
    replication_group_id = "${concat(var.epop, \"-redis\")}"
    description = "test-sjc-redis"
    engine = "redis"
    cache_node_type = "cache.m3.medium"
    num_cache_clusters = 2
    automatic_failover = true
    subnet_group_name = "${aws_elasticache_subnet_group.redis.name}"
    security_group_ids = ["${aws_security_group.redis.id}"]

    preferred_cache_cluster_azs = [
        "us-west-1a",
        "us-west-1c",
    ]

    depends_on = [
        "aws_elasticache_subnet_group.redis",
        "aws_security_group.redis"
    ]
}
```

**Comment by catsby** [comment]
Hey @saulshanabrook thank you for the code contribution! Would you by chance be willing to add documentation for this new resource while I review the code and try it out? 

Thanks!

**Comment by catsby** [comment]
Am I correct in that this (as it stands) is not meant to create a cluster replication group for an existing ElastiCache Redis Cluster? There's no mention or use of `PrimaryClusterId` here. It seems that attribute is optional if `NumCacheClusters` is specified, but I'm wondering if we could add that here and support both cases?

File: builtin/providers/aws/resource_aws_elasticache_replication_group.go [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+	_, err := conn.CreateReplicationGroup(req)
+	if err != nil {
+		return fmt.Errorf("Error creating Elasticache replication group: %s", err)
+	}
+
+	pending := []string{"creating"}
+	stateConf := &resource.StateChangeConf{
+		Pending:    pending,
+		Target:     "available",
+		Refresh:    ReplicationGroupStateRefreshFunc(conn, d.Id(), "available", pending),
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_elasticache_replication_group.go]: `ReplicationGroupStateRefreshFunc` should not be exported, right? `replicationGroupStateRefreshFunc` instead.

Also, you send `d.Id()` as the id to `ReplicationGroupStateRefreshFunc` , before setting it with `SetId`, thus sending `""`. This causes the function to return false positives in the event that you already have a replication group defined `""` will return all, and then you grab the first. In my case, I had two defined, so I got a false positive that my newest one was up and available. 

Calling `SetId` before making the `WaitForState()` call should fix this :smile:

File: builtin/providers/aws/resource_aws_elasticache_replication_group.go [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+		if ec2err, ok := err.(awserr.Error); ok && ec2err.Code() == "ReplicationGroupNotFoundFault" {
+			// Update state to indicate the replication group no longer exists.
+			d.SetId("")
+			return nil
+		}
+
+		return err
+	}
+
+	if len(res.ReplicationGroups) == 1 {
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_elasticache_replication_group.go]: This block will cause a panic if `ReplicationGroups[0]` is in the `deleting` state (none of those attributes are set, so you get `nil` reference issues). Also, why do we set `primary_endpoint` outside of this block?

**Comment by saulshanabrook** [comment]
@catsby Yeah it looks like there are two ways to create a Replication Group and we should support both. 

from [the API docs](http://docs.aws.amazon.com/AmazonElastiCache/latest/APIReference/API_CreateReplicationGroup.html): 

> PrimaryClusterId
The identifier of the cache cluster that will serve as the primary for this replication group. This cache cluster must already exist and have a status of available.
>
> This parameter is not required if NumCacheClusters is specified.

Also I will add a `ReplicationGroupId` field to the `ElasticacheCluster`, which allows you to add individual clusters as read replicas.

File: builtin/providers/aws/resource_aws_elasticache_replication_group.go (hunk 1 above)
Comment (saulshanabrook) [inline builtin/providers/aws/resource_aws_elasticache_replication_group.go]: Good catch, just fixed, thank you!

File: builtin/providers/aws/resource_aws_elasticache_replication_group.go (hunk 2 above)
Comment (saulshanabrook) [inline builtin/providers/aws/resource_aws_elasticache_replication_group.go]: How do I check the state?

Comment (catsby) [inline builtin/providers/aws/resource_aws_elasticache_replication_group.go]: The ReplicationGroup status is in the output struct of the `DescribeReplicationGroups` method

- https://github.com/aws/aws-sdk-go/blob/master/service/elasticache/api.go#L4756

If I recall, the issue appeared when you reference `res.ReplicationGroups[0].NodeGroups[0].PrimaryEndpoint.Address`... but it could also be that some of the other elements simply weren't present. I was only able to trigger it by manually deleting the group in the console, then running `terraform refresh`

**Comment by catsby** [comment]
Hey @saulshanabrook – sorry for the silence, is this ready for another review? Thanks again!

**Comment by saulshanabrook** [comment]
@catsby Yep

**Comment by saulshanabrook** [comment]
@catsby Anything else that I should do on my end?

**Comment by joshgarnett** [comment]
Any ETA on this?

**Comment by munhitsu** [comment]
+1

**Comment by nicolai86** [comment]
Would love this to get merged...

**Comment by dpetzold** [comment]
+1

**Comment by dpetzold** [comment]
fyi I had to make the following changes to get it to work with master

https://github.com/clearcare/terraform/commit/2c02b57173d5f7dad604ee4175ca1087b8e2777b

**Comment by saulshanabrook** [comment]
@dpetzold I cherry picked your commit and rebased against master.

**Comment by gposton** [comment]
+1 on merging... we're running a forked version of terraform for this and would love to get back on 'main'

File: website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+Either this or `primary_cluster_id` is required.
+
+* `primary_cluster_id` - (Optional) The identifier of the cache cluster that
+will serve as the primary for this replication group. This cache cluster must already exist and have a status of available.
+Either this or `num_cache_clusters` is required.
+
+* `parameter_group_name` – (Required) Name of the parameter group to associate
+with this cache cluster
+
+* `preferred_cache_cluster_azs` - (Optional) A list of EC2 availability zones in which the replication group's cache clusters will be created. The order of the availability zones in the list is not important. If not provided, AWS will chose them for you.
```
Comment (thegedge) [inline website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown]: s/chose/choose/

File: website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+The following arguments are supported:
+
+* `replication_group_id` – (Required) Replication group identifier. This
+parameter is stored as a lowercase string
+
+* `description` – (Required) The description of the replication group.
+
+* `engine` – (Optional) The name of the cache engine to be used for the cache clusters in this replication group.
+ The only current valid value is `redis`
```
Comment (thegedge) [inline website/source/docs/providers/aws/r/elasticache_replication_group.html.markdown]: I'd reword the last sentence to "The only value that is currently valid is `redis`"

File: builtin/providers/aws/resource_aws_elasticache_replication_group.go [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"preferred_cache_cluster_azs": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				ForceNew: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
+			},
```
Comment (thegedge) [inline builtin/providers/aws/resource_aws_elasticache_replication_group.go]: Perhaps we should add some of the other parameters – like snapshot info, preferred maintenance window, and anything else in [CreateReplicationGroupInput](http://docs.aws.amazon.com/sdk-for-go/api/service/elasticache.html#type-CreateReplicationGroupInput) that currently isn't here?

**Comment by thegedge** [comment]
Was just about to sit down and whip up a PR to do exactly this, so if there's anything we can do to help move this along to getting merged, let me know.

**Comment by saulshanabrook** [comment]
@thegedge Awesome, thanks for taking a look at this. PR's against my branch would be really great, so that I can add your changes easier.

File: builtin/providers/aws/resource_aws_elasticache_replication_group.go (hunk 5 above)
Comment (saulshanabrook) [inline builtin/providers/aws/resource_aws_elasticache_replication_group.go]: @thegedge That makes sense to me.

**Comment by jamesob** [comment]
+1. Would like to have this in.

**Comment by FlorinAndrei** [comment]
This would be incredibly useful, please merge.

**Comment by jgross206** [comment]
+1 would love this

**Comment by apparentlymart** [comment]
For various other ElastiCache resources we had issues where the AWS API would accept mixed-case names for various things but then normalize them to lowercase and fail if we try to retrieve them using the original mixed-case names. See for example #3235 and #3120.

To make sure a similar bug isn't lurking here it'd be good to make the `replication_group_id` in the acceptance test include some uppercase characters. If that makes it fail, then the fix is probably similar to that of the two issues I linked above. (Sorry I don't have my Terraform dev env handy to test it myself right now.)

Separately, I have a terminology question: for several of the other ElastiCache objects the main identifier for the resource is `name`. In this case it's called `replication_group_id`. Would it be appropriate to rename the argument to `name` for consistency with the others, or is "Replication Group Id" how this is referred to in the AWS console UI?

**Comment by saulshanabrook** [comment]
@apparentlymart That sounds like a good idea, for the case normalization. I am not working actively on this at the moment, but am happy to accept PR's against my branch.

**Comment by HarryAtMove** [comment]
Would like to be using this, please merge

File: .gitignore [hunk 6]
```diff
@@ -18,3 +18,4 @@ website/node_modules
 *.bak
 *~
 .*.swp
+config/lang/y.go
```
Comment (catsby) [inline .gitignore]: This should not be in the `.gitignore`.  There's some oddness with go 1.4 and 1.5 but for now this file should just not be included in any changes, more ignored

File: builtin/providers/aws/resource_aws_elasticache_cluster.go [hunk 7]
```diff
@@ -72,6 +72,11 @@ func resourceAwsElasticacheCluster() *schema.Resource {
 				Computed: true,
 				ForceNew: true,
 			},
+			"replication_group_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_elasticache_cluster.go]: This seems like a `Computed` attribute to me, what do you think? It seems we would have a `aws_elasticache_cluster` resource, and then a `elasticache_replication_group` which has a `primary_cluster_id` that points to the `aws_elasticache_cluster`. The `aws_elasticache_cluster` can exist by itself, but it may have a replication group associated with it, which comes later. Does that make sense?

File: builtin/providers/aws/resource_aws_elasticache_replication_group.go [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+		AutomaticFailoverEnabled:    aws.Bool(automaticFailover),
+		NumCacheClusters:            aws.Int64(int64(numCacheClusters)),
+		PrimaryClusterId:            aws.String(primaryClusterID),
+		Engine:                      aws.String(engine),
+		CacheSubnetGroupName:        aws.String(subnetGroupName),
+		EngineVersion:               aws.String(engineVersion),
+		CacheSecurityGroupNames:     securityNames,
+		SecurityGroupIds:            securityIds,
+		PreferredCacheClusterAZs:    prefferedAZs,
+	}
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_elasticache_replication_group.go]: if any of these above attributes are omitted because they are optional, this `CreateReplicationGroupInput` will contain empty strings, which is typically not accepted in the API. Does this work when they are omitted?

Comment (catsby) [inline builtin/providers/aws/resource_aws_elasticache_replication_group.go]: Discovered on my own that the answer is "no", but we can fix this easily :smile: Looks like `engine_version`, `preferred_cache_cluster_azs` and `security_group_names` specifically need to be conditionally added

**Comment by catsby** [comment]
I posted a few notes, but I think this needs some follow up. We need to merge/rebase master, and we need to evaluate what we're sending to the API. I got many "invalid parameter combination" errors when trying to specify all/some of the attributes.

**Comment by saulshanabrook** [comment]
@catsby I am not actively working on this, at this time. Anyone else is welcome to follow up on these changes.

**Comment by dbethke** [comment]
+1. I'd like to use this.

**Comment by saulshanabrook** [comment]
I am going to close this for now, since I am not going to fix it up.

**Comment by grayaii** [comment]
Don't forget about tags... We're building from @dpetzold  branch, so if possible we'll update it there...

**Comment by woodhull** [comment]
@catsby this seems like a feature that hashicorp should invest in given the original submitter closed the issue... and there are lots of people running a forked version to get the functionality.

**Comment by chadgrant** [comment]
@catsby Need this also

**Comment by aldarund** [comment]
+1, need this too

**Comment by HarryAtMove** [comment]
we could use this as well

**Comment by danielnbarbosa** [comment]
+1 this would be great to have

**Comment by dlsniper** [comment]
Can someone please lock this CLOSED issue so that it doesn't generate anymore +1 e-mails? Thank you!

**Comment by gposton** [comment]
@catsby @dlsniper, @dpetzold and I intend to follow up on the work that @saulshanabrook did here.  From the inline comments, it doesn't look like there's much left to do.  Am I missing something?

**Comment by inferiorhumanorgans** [comment]
@catsby any update on this?

**Comment by apparentlymart** [comment]
I've locked this since the original submitter has said he no longer intends to work on this and so further discussion here isn't productive.

If someone would like to build on this and submit a new PR then that would be much appreciated!

**Comment by catsby** [comment]
Hey Friends – sorry for the silence here. @stack72 implemented `aws_elasticache_replication_group ` in https://github.com/hashicorp/terraform/pull/8275, check it out!  If you find something missing or have other feedback, please open a new issue

Thanks!
